trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java

```diff
@@ -0,0 +1,471 @@
+					mLogger.error("unable to determine which branch was taken, unable to determine the location");
+				} else {
+					mLogger.warn("unable to determine which branch was taken at " + loc);
+				}
+				return;
+			}
+			for (final Entry<TermVariable, CodeBlock> entry : bi2cb.entrySet()) {
+				final boolean taken = branchEncoders.get(entry.getKey());
+				if (taken) {
+					addCodeBlock(entry.getValue(), relevanceInformation, threadId, forkedThreadId, joinedThreadId,
+							trace, branchEncoders);
+					return;
+				}
+			}
+			throw new AssertionError("no branch was taken");
+		} else if (cb instanceof GotoEdge) {
+			// skip goto edges
+			return;
+		} else {
+			throw new UnsupportedOperationException("Unsupported CodeBlock: " + cb.getClass().getCanonicalName());
+		}
+		trace.add(ateBuilder.build());
+	}
+
+	@Override
+	public IProgramExecution<BoogieASTNode, Expression>
+
+			translateProgramExecution(final IProgramExecution<IIcfgTransition<IcfgLocation>, Term> programExecution) {
+		if (!(programExecution instanceof IcfgProgramExecution)) {
+			throw new IllegalArgumentException();
+		}
+		final IcfgProgramExecution icfgProgramExecution = (IcfgProgramExecution) programExecution;
+		assert checkCallStackSourceProgramExecution(mLogger,
+				programExecution) : "callstack of initial program execution already broken";
+
+		final List<AtomicTraceElement<BoogieASTNode>> trace = new ArrayList<>();
+		final Map<Integer, ProgramState<Expression>> programStateMapping = new HashMap<>();
+
+		if (icfgProgramExecution.getInitialProgramState() != null) {
+			programStateMapping.put(-1, translateProgramState(icfgProgramExecution.getInitialProgramState()));
+		}
+		for (int i = 0; i < icfgProgramExecution.getLength(); i++) {
+			final AtomicTraceElement<IIcfgTransition<IcfgLocation>> ate = icfgProgramExecution.getTraceElement(i);
+			final Map<TermVariable, Boolean>[] branchEncoders = icfgProgramExecution.getBranchEncoders();
+			if (branchEncoders == null || i >= branchEncoders.length) {
+				addCodeBlock(ate, trace, null);
+			} else {
+				addCodeBlock(ate, trace, branchEncoders[i]);
+			}
+			final int posInNewTrace = trace.size() - 1;
+			final ProgramState<Term> programState = icfgProgramExecution.getProgramState(i);
+			programStateMapping.put(posInNewTrace, translateProgramState(programState));
+		}
+		assert checkCallStackTarget(mLogger, trace);
+		return new BoogieProgramExecution(programStateMapping, trace, programExecution.isConcurrent());
+	}
+
+	@Override
+	public IBacktranslatedCFG<String, BoogieASTNode>
+			translateCFG(final IBacktranslatedCFG<IcfgLocation, IIcfgTransition<IcfgLocation>> cfg) {
+		final IBacktranslatedCFG<String, BoogieASTNode> translatedCfg =
+				translateCFG(cfg, (a, b, c) -> translateCFGEdge(a, (IIcfgTransition<IcfgLocation>) b, c));
+		// mLogger.info(getClass().getSimpleName());
+		// printHondas(cfg, mLogger::info);
+		// printCFG(cfg, mLogger::info);
+		// mLogger.info("######## END "+getClass().getSimpleName());
```

Comment:
Remove unused debugging code or implement it correctly when debug mode is enabled.